dfdaemon/config/config.go

```diff
@@ -67,6 +79,112 @@ type Properties struct {
 	// are provided, all requests will be proxied directly. Request will be
 	// proxied with the first matching rule.
 	Proxies []*Proxy `yaml:"proxies"`
+
+	// HijackHTTPS is the list of hosts whose https requests should be hijacked
+	// by dfdaemon. Dfdaemon will be able to proxy requests from them with dfget
+	// if the url matches the proxy rules
+	HijackHTTPS *HijackConfig `yaml:"hijack_https"`
+}
+
+// HijackConfig represents how dfdaemon hijacks http requests
+type HijackConfig struct {
+	Cert  string        `yaml:"cert"`
+	Key   string        `yaml:"key"`
+	Hosts []*HijackHost `yaml:"hosts"`
+}
+
+// HijackHost is a hijack rule for the hosts that matches Regx
+type HijackHost struct {
+	Regx  *Regexp   `yaml:"regx"`
+	Certs *CertPool `yaml:"certs"`
+}
+
+// CertPool is a wrapper around x509.CertPool, which can be unmarshalled and
+// constructed from a list of filenames
+type CertPool struct {
+	files []string
+	*x509.CertPool
+}
+
+// UnmarshalYAML implements yaml.Unmarshaller
+func (cp *CertPool) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	return cp.unmarshal(unmarshal)
+}
+
+// UnmarshalJSON implements json.Unmarshaller
+func (cp *CertPool) UnmarshalJSON(b []byte) error {
+	return cp.unmarshal(func(v interface{}) error { return json.Unmarshal(b, v) })
+}
+
+func (cp *CertPool) unmarshal(unmarshal func(interface{}) error) error {
+	if err := unmarshal(&cp.files); err != nil {
+		return err
+	}
+
+	pool, err := certPoolFromFiles(cp.files...)
+	if err != nil {
+		return err
+	}
+
+	cp.CertPool = pool
+	return nil
+}
+
+// Regexp is simple wrapper around regexp.Regexp to make it unmarshallable from a string
+type Regexp struct {
+	*regexp.Regexp
+}
+
+// NewRegexp returns new Regexp instance compiled from the given string
+func NewRegexp(exp string) (*Regexp, error) {
+	r, err := regexp.Compile(exp)
+	if err != nil {
+		return nil, err
+	}
+	return &Regexp{r}, nil
+}
+
+// UnmarshalYAML implements yaml.Unmarshaller
+func (r *Regexp) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	return r.unmarshal(unmarshal)
+}
+
+// UnmarshalJSON implements json.Unmarshaller
+func (r *Regexp) UnmarshalJSON(b []byte) error {
+	return r.unmarshal(func(v interface{}) error { return json.Unmarshal(b, v) })
+}
+
+func (r *Regexp) unmarshal(unmarshal func(interface{}) error) error {
+	var s string
+	if err := unmarshal(&s); err != nil {
+		return err
+	}
+	exp, err := regexp.Compile(s)
+	if err == nil {
+		r.Regexp = exp
+	}
+	return err
+}
+
+// MarshalJSON implements json.Marshaller to print the regexp
+func (r *Regexp) MarshalJSON() ([]byte, error) {
+	return json.Marshal(r.String())
+}
+
+// certPoolFromFiles returns an *x509.CertPool constructed from the given files.
+// If no files are given, (nil, nil) will be returned.
```

Comment:
// If no files are given, an empty CertPool will be returned.

Reply:
My mistake. I meant to return `(nil, nil)` when `len(files) == 0` so it will be easier to tell if a cert pool is configured. I'll fix it right away. Checking the pointer is safer that checking the nested field.

Reply:
My mistake. I meant to return `(nil, nil)` when no files are given. It's easier to tell if the cert pool is configured. I'll add a `len(files) == 0` check at the beginning of the function.

And you remind me that if `certs` is an empty list, we should check for the nested `CertPool` as well.